### ISSUE TITLE
Running out of disk space on macos

### ISSUE THREAD

**Issue by sramam** [pnpm/pnpm#6880]
I am using pnpm to save disk space on a macbook air. 
But it seems the macos disk utility double counts the hardlinks setup by pnpm. 
This results in the system running out of disk space, despite have 10s of GB likely free.

I keep running into disk space issues within other tools not having space on disk to write. 
Deleting the node_modules in this picture adds that 474MB back to the free disk space number.

I have a lot of pnpm repos so this is starting to become a real bother. 

> Is best practice to set the virtual-store-dir to a common point to avoid this problem?

<img width="543" alt="image" src="https://github.com/pnpm/pnpm/assets/518574/2ea45860-4fc1-4ef9-8cd1-ed72abb6c8ed">

### pnpm version:
8.6.9

### Expected behavior:
I'd like to be able to use my disk without double-counting 

### Actual behavior:
Disk space double counted and acting as if "occupied"

### Additional information:

❯ node -v
v18.13.0
macos 12.4 MBA-m2

**Comment by zkochan** [comment]
> Is best practice to set the virtual-store-dir to a common point to avoid this problem?

no, the virtual store cannot be shared between projects.

Are you sure hard links are created? It doesn't make sense that macOS would count the disk space incorrectly. Try to set the `package-import-method` to `hardlink` explicitly (https://pnpm.io/npmrc#package-import-method)

**Comment by sramam** [comment]
Based on this node, I presumed it was using hard-links.

> INFO
> Even though macOS supports cloning, there is currently [a bug in Node.js](https://github.com/libuv/libuv/pull/2578) that prevents us from using it in pnpm. If you have ideas how to fix it, [help us](https://github.com/pnpm/pnpm/issues/5001).

Now that I am digging in, between the link you posted and the node_linker behavior, 
I am unsure what pnpm is doing.

I was just reading the label, attracted by the promise of disk space savings and using it.
Are there recommended settings different from default to save disk-space on macos?
Pointers to documentation would be appreciated.

**Comment by zkochan** [comment]
Yes, it should use hardlinks by default on macOS. But it will fall back to other means if it cannot create a hardlink. That is why I suggested to explicitly set it.

There are no other settings related to it. By default it should work.

**Comment by zkochan** [comment]
Closing: this is not something pnpm can fix.

Finder counts each hard link's bytes again, so the same files are reported several times. The bytes on disk are counted once.

If you still hit this on the latest pnpm, please comment with a reproduction and we will reopen.

---
Written by an agent (Claude Code, claude-opus-5).
